doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+# RFC 001: Extend Quint with Row-Polymorphic Sum Types
+
+**Table of Contents**
+
+- [tl;dr](#tldr)
+- [Motivation](#motivation)
+- [Context](#context)
+- [Statics](#statics-1)
+- [Dynamics](#dynamics-1)
+- [Concrete Syntax](#concrete-syntax)
+- [Additional consideration](#additional-consideration)
+
+## tl;dr
+
+This section gives a concise overview of the principle proposal. In depth
+discussion follows in the subsequent sections.
+
+### Concrete syntax
+
+Here I illustrate two attractive alternatives for each critical construct. See
+[Syntax](#syntax) for a discussion of motivations and trade offs.
+
+#### Declaration
+
+This group of alternatives is closer to Rust, but with significant variation for
+reasons discussed in [Syntax](#syntax).
+
+```quint
+type T =
+  | A(int)
+  | B(str)
+```
+
+Variants labels with no associated expression type are allowed, but these are only
+sugar for a label typed with the unit (i.e., the empty record):
+
+```quint
+type S = C
+```
+
+Is normalized as
+
+```quint
+type S = C({})
+```
+
+
+#### Injection
+
+I propose introducing a lexical rule to syntactically differentiate data
+constructor of sum types from operator application: any identifiers starting
+with an uppercase letter that is occurs in the syntactic position of operator
+application is parsed as a sum type constructor: 
+
+```quint
+// Constructoin of a previously declared sum type
+val a : T = A(1)
+// Construction of a polymorphic variant of an anonymous type
+val b : AdHocVariant(int) | v = AdHocVariant(1)
+```
+
+Identifiers starting with a capital letter would them be available only for
+module names, type aliases, and variant constructors.
```

Comment:
This is probably too restrictive. When people are specifying protocols that come from the Golang or Rust implementation, they might want to keep they uppercase notation. I have seen it several times already.

Reply:
Yeah, they’re tradeoffs here either way.

Reply:
See https://github.com/informalsystems/quint/issues/1081 for discussion on this point. I've removed this part from the proposal.